### PR TITLE
Improve magic display system

### PR DESCRIPTION
**Pull request by eranwitkon**
I have changed the InterpreterResult behavior so that if a magic word is detected all preceding text is omitted. This way we can use the display system and print out using the magic words in paragraphs such as spark.
example:
```
sc.parallelize(1 to 10)
```
will output:
```
res9: org.apache.spark.rdd.RDD[Int] = ParallelCollectionRDD[2] at parallelize at <console>:24
```  

and mix paragraphs such as:
```
sc.parallelize(1 to 10)
println("%table col\tcol2\nqwe\t123\n")
```
will output the expected magic table
selection_003

This behavior can be later extended to handle complex types.
I have also added a JUnit test for the class

### CODE REVIEW

**Comment by doanduyhai** [comment]
A question. What if I mix different magics ? Example:

```scala
sc.parallelize(1 to 10)
println("%table col\tcol2\nqwe\t123\n")
println("%html <h3> This is a hack </h3>")
```

 In that case, what would be displayed ? A table or the HTML bloc ? I guess that in the source code of method `getType()` when we loop through all the types enums, the first type found in the output string will be retained.

But then doing `msg.substring(magicLength + msg.indexOf(magic))` will give very different results depending on whether we match **%table%** or **%html%** before.

Is this behavior deterministic or just depending on the order of the type enums ?

**Comment by goi** [comment]
Yes the current behavior is depending on the order of the type enums, and a better solution is to allow complex output. BUT I argue that the current implementation is not deterministic as well as the output of the magic word depends on the previous line of code and if the underling engine returned some text before.
If it is important I can change the loop to track the last occurrence of the magic word and trim all preceding text  that way in your example the %html will be presented.

**Comment by doanduyhai** [comment]
I guess that a rule saying: **_if you mix several magics in the same output message, only the last one will be taken into account_** can be a good start for this PR.

This way, whatever the users do, they'll always have a **consistent** and **predictable** behavior.

To implement this, maybe in the for loop, for each matching magic, you may save:

1. the magic word
2. the index of this magic word

And after the for loop, just take the magic with highest index and strip out all the text before. 

How does it sound ?

**Comment by goi** [comment]
OK That's sounds right.
I will update the code and commit again.
Thanks,
Eran

**Comment by goi** [comment]
I have updated the code to comply with this rule:
if you mix several magics in the same output message, only the last one will be taken into account
This include the following cases:
* text return text
* text before magic return magic
* magic A before magic B return magic B
* magic A before magic A' return magic A'

Updated include unit test and update to the POM
See example output:
selection_004

**Comment by doanduyhai** [comment]
Looks good to me. One possible refactoring to avoid the duplication of output message parsing:

```java
private TreeMap<Integer, Type> buildIndexMap(String msg) {
    TreeMap<Integer, Type> typesLastIndexInMsg = new TreeMap<Integer, Type>();
    if (msg.startsWith(magic + " ") || msg.startsWith(magic + "\n")) {
        for (Type t : types){
            lastIndexOftypes = getLastIndexOfType(msg, t);
            if (lastIndexOftypes >= 0){
                typesLastIndexInMsg.put(lastIndexOftypes, t);
            }
        }
    }
    return typesLastIndexInMsg;		    
}
```

And then re-use this method in `getType()` and  `getData()`

**Comment by goi** [comment]
I have refactor the code as suggested. just one note, I dropped the line:
```
if (msg.startsWith(magic + " ") || msg.startsWith(magic + "\n")) {
```
in the suggested buildIndexMap as my understanding is that the msg can start with any text not just magic word.
All tests passed successfully.

**Comment by doanduyhai** [comment]
Perfect, it's ready to be merged then

File: zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/InterpreterResult.java [hunk 1]
```diff
@@ -87,36 +89,51 @@ private String getData(String msg) {
...
-        }
-      }
+    Type[] types = type.values();
+    TreeMap<Integer, Type> typesLastIndexInMsg = buildIndexMap(msg);
+    if (typesLastIndexInMsg.size() == 0) {
+      return msg;
+    } else {
+      Map.Entry<Integer, Type> lastType = typesLastIndexInMsg.lastEntry();
+      int magicLength = lastType.getValue().name().length() + 1; //add 1 for the %
+      int subStringPos = magicLength + lastType.getKey() + 1; // 1 for the last \w after magic
```
Comment (felixcheung) [inline zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/InterpreterResult.java]: \w? do you mean newline or regex:\w?
if latter, what if someone types "%" - ie. no character following %

Comment (eranwitkon) [inline zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/InterpreterResult.java]: Thanks for the question.
1) I fixed the comments to be more clear
2) a msg with only % will not get to that point as if
(typesLastIndexInMsg.size() == 0) which means we return  return msg;
3) added test cases for this in unit test

Eran

Comment (felixcheung) [inline zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/InterpreterResult.java]: Cool!!

Comment (eranwitkon) [inline zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/InterpreterResult.java]: Can someone look at this one and merge it if possible?
Eran

בתאריך שבת, 25 ביולי 2015, 22:51 מאת Felix Cheung <notifications@github.com

**Comment by Leemoonsoo** [comment]
Thanks for nice improvement. LGTM. While it is changing end-user experience (output handled differently in certain case), I'd like to let this open for next 24 hours and merging it if there're no more discussions.

**Comment by Leemoonsoo** [comment]
Just before merge, i have tested one more.

There were feature that display html code inside of cell in table, (Section 'Display as Table, chart' http://zeppelin.incubator.apache.org/docs/display.html)

for example
```
println(s"""%table
key\tvalue
a\t%html <font color=red>red</font>
""")
```

It supposed to be displayed in this way
image

but displayed like this in this branch.
image

**Comment by eranwitkon** [comment]
I see, this changes the rule we thought of but it needs to be fixed.
I suggest the new rule: paragraph type will be set by the first magic word
in the paragraph. This will solve the example but in case %html some text
%table it will not process the %table.
Or do you have other suggestion for the desire outcome?
BTW, I will be oof till Wednesday so will fix it late next week.
Eran

בתאריך שבת, 1 באוג׳ 2015, 18:54 מאת Lee moon soo <notifications@github.com>:

**Comment by eranwitkon** [comment]
I changed the implementation according to the new rule:
"paragraph type will be set by the first magic word in the paragraph" this allows for test such as the following to be parsed as before 
```
println(s"some text before magic word %table col1\tcol2\naaa\t%html <h3>123</h3>\n")
```
selection_005

also support omitting text before magic word so magic can be used in spark code
```
val somerdd = sc.parallelize(1 to 10)
println("%table key\tvalue\nrdd count   \t%html <h3>" + somerdd.count +"</h3>\n")
```
selection_006

one thing to note is that it is up to the first magic word to handle the proceeding text. e.g. %table knows how to pars %html within it. but %table followed by %table will be handled differently.
```
println("%table col A\tcol B\naaa\t123\n")
println("%table col Z\naaa\n123\n")
```
selection_007

if no more changes, this is ready for merge
Eran

**Comment by elbamos** [comment]
Perhaps it makes sense to start moving to the point where interpret returns a collection of InterpreterResults rather than just one? Or that we allow one type of InterpreterResult to contain multiple nested InterpreterResults?

**Comment by eranwitkon** [comment]
I agree that in the long run this could be a better solution but I believe
that making small steps here is important and solve current problems.
For my users using this change is valuable by itself.
I am willing to see if I can do the next step as well once we have this one
completed.
Eran

**Comment by felixcheung** [comment]
I guess maybe a list of InterpreterResult is where we should get to, later? so then we will have one InterpreterResult per unit of code or something.

The current approach here makes sense to me.

**Comment by Leemoonsoo** [comment]
LGTM

Also +1 for having collection of InterpreterResult as a next step.

**Comment by felixcheung** [comment]
Created https://issues.apache.org/jira/browse/ZEPPELIN-212

**Comment by eranwitkon** [comment]
I hope we are going to merge this one before moving to the list
implementation. Eran

בתאריך יום א׳, 9 באוג׳ 2015, 20:36 מאת Felix Cheung <
notifications@github.com>:

**Pull request closed without merging**